### ISSUE TITLE
Implementation options for Generators

### ISSUE THREAD

**Issue by fatcerberus** [svaarala/duktape#1101]
It occurs to me that Duktape probably already has most of the plumbing in place to support generators, as part of the implementation for Duktape threads.  What, concretely, is needed to extend that to support ES6 generator functions?

**Comment by svaarala** [comment]
Not sure, haven't given it much thought yet - but I'd like to ensure overlap between Duktape threads and GeneratorFunctions would be minimized. For example, maybe Duktape threads could be special kinds of GeneratorFunction instances.

This issue is a duplicate of #1012 (or vice versa :).

**Comment by fatcerberus** [comment]
Ah, didn't realize there was an issue for this already, sorry for the noise :)

**Comment by fatcerberus** [comment]
One preliminary change that I think will be necessary is to relax the "no native calls on the stack" restriction.  In practice that likely means yield/resume must be implemented using something other than `longjmp()`.

**Comment by svaarala** [comment]
That's a long standing issue and not easy to solve portably. Switching stacks is very unportable, and requiring native functions to participate in resuming from a yield later is somewhat awkward and only works if all native calls involved participate in resuming. So that should be addressed separately I think; maybe the solution is stack switching but that should then be optional because it's a major portability hassle.

But to be really pedantic, the specification cannot require that *any* native calls be allowed because the behavior of native calls is not really specified (after all, there's no way to declare one). But what *is* required is that all built-ins (native or not) and Ecmascript functions are allowed.

**Comment by fatcerberus** [comment]
Note I said *relax*, not necessarily eliminate.  I think at present it's impossible to resume if there's a native call anywhere on the callstack (correct me if I'm wrong ;); I don't consider it unreasonable at all that one can't resume *across* a native call, but if all the yielding and resuming is happening *above* such a call there shouldn't be an issue?

Anyway the reason I brought this up is that it might be a bit surprising for script writers to call a generator and then immediately have `next()` throw an error because there happened to be a native call at some lower part of the callstack.  It's not a spec concern but a practical one. :)

**Comment by svaarala** [comment]
There cannot be native calls in the callstack of the yielding thread (because those couldn't be recreated on resume without stack switching or similar). There are no limitations on the thread doing the resume call.

**Comment by svaarala** [comment]
So to be clear, there would be no limitations on calling `.next()` with whatever calls in the callstack of the thread calling `.next()`. That's already the case now for coroutines. The limitations are for the yielding thread (= the thread that is resumed).

**Comment by fatcerberus** [comment]
Ah, then ignore what I said above then, I was under the mistaken impression that one couldn't resume with native calls in progress.  If it's only that you can't yield then that's immaterial for generators - since I think only the generator function *itself* can yield anyway.

**Comment by svaarala** [comment]
What would be nice is if Duktape threads would be "extended generators" that *could* yield from other places too but otherwise work like ES6 generators to the caller. This can't be done via syntax but via `Duktape.Thread.yield()` (perhaps moved to some more logical place).

**Comment by fatcerberus** [comment]
(Thinking out loud a bit here, don't mind me :)

Hm, so with the above in mind, ES6 generators could conceivably be implemented as syntactic sugar for what is now done through the `Duktape.Thread` namespace.  I would have to read the spec (of course) to verify all the details, but from a cursory look at the basic generator setup, it seems that calling a generator function is like calling `Thread.create()` while `gen.next()` is like calling `Thread.resume(thr)`.  The APIs align pretty nicely, it seems.

Maybe I'll try experimenting with this soon.  It would be a useful feature to have for my `from.js` query library.  It's currently pretty basic and can't, e.g. sort results *in situ* or synthesize new results (like e.g. SelectMany in LINQ).  Generators would greatly simplify the implementation of those features. (I can't use Duktape threads because I need my standard library to be portable ;-)

**Comment by svaarala** [comment]
It won't be just syntactic sugar because generator functions have special methods like `.next()`. Then again, if Duktape.Thread.prototype had a `.next()` it'd be pretty close.

**Comment by svaarala** [comment]
Hmm, so if you wanted to hack around:
- Add a `.next()` binding for Duktape.Thread instances via `Duktape.Thread.prototype`. Roughly the same functionality as a resume() except you can't inject errors.
- Compile a yield into the equivalent of `Duktape.Thread.yield()`. Perhaps needs a separate opcode because it needs to access the original `Duktape.Thread` (so it needs to go through `thr->builtins[]`).

I'm not sure if this would be even close to an eventual solution - like I said, I'd like to figure out how to minimize mechanisms when aligning with GeneratorFunctions. Hacking would still be valuable to gain some insight of what changes are needed in more detail.

**Comment by svaarala** [comment]
For example one thing that I'd like to support even later on is the ability to resume-with-error and yield-an-error as they are quite useful in some situations. They can be handled via instance methods, e.g. `foo.nextError()` and `foo.yieldError()` if there's no cleaner way.

**Comment by fatcerberus** [comment]
Actually it's not the generator function that has a `.next()` method but the generator *instances* you get by calling it.  That is, the following are roughly equivalent:

With Duktape threads:
```js
function genNumbers()
{
    var i = 0;
    while (true) { Duktape.Thread.yield(++i); }
}

var thr = new Duktape.Thread(genNumbers);
console.log(Duktape.Thread.resume(thr));  // 1
console.log(Duktape.Thread.resume(thr));  // 2
console.log(Duktape.Thread.resume(thr));  // 3
```

With ES6 generators:
```js
function* genNumbers()
{
    var i = 0;
    while (true) { yield ++i; }
}

var genInst = genNumbers();
console.log(genInst.next());  // 1
console.log(genInst.next());  // 2
console.log(genInst.next());  // 3
```

Pretty much a 1:1 mapping, the only exception is that the ES6 generator function can take arguments, which doesn't seem to be possible with Duktape threads.

**Comment by svaarala** [comment]
Right, that's what I meant - generator instances would map to Duktape.Thread instances.

**Comment by fatcerberus** [comment]
I made this adapter that converts a normal Ecmascript function into an ES6 generator (save for the `yield` keyword):

```js
function makeGen(fn) {
	return function()
	{
		var thr = new Duktape.Thread(fn);
		return { next: next };
		
		function next()
		{
			var value = Duktape.Thread.resume(thr);
			return { value: value, done: false };
		}
	};
}
```

However, there doesn't seem to be a way to determine whether the function returned (as opposed to yielding)?  Thus I can never set `done` to true and the next call to `.next()` throws an error because the coroutine is dead.

**Comment by svaarala** [comment]
>However, there doesn't seem to be a way to determine whether the function returned (as opposed to yielding)? Thus I can never set done to true and the next call to .next() throws an error because the coroutine is dead.

There's internal book-keeping for that so exposing it in some way shouldn't be too difficult. That book-keeping is actually a bit annoying (it interferes with some other goals) so I've been looking into removing or reworking it. But maybe I could expose it via Duktape.info()?

**Comment by fatcerberus** [comment]
It's kind of annoying that the state isn't exposed because resume() will throw if the thread has died.  ES6 generators will just return undefined if you resume them after they've finished.

It does me good to catch the error either, because to fully emulate ES6 I have to set `done` to true on the final yield (i.e. the `return`), not the call afterwards.

Anyway, the `Duktape.info()` change would work :)

**Comment by svaarala** [comment]
Yep, but quite naturally there are dozens of internal state fields that are intentionally not exposed unless they're really needed. Doing so limits the fallout from changing internals. So while it's annoying in every particular case, it's still a good basic policy as far as information hiding is concerned.

**Comment by svaarala** [comment]
Changing Duktape.Thread to return undefined when a finished coroutine is resumed would be a trivial change and maybe sensible if it more closely matches ES6 behavior.

**Comment by fatcerberus** [comment]
That'd probably be a good change, however it wouldn't be enough to know for sure if the thread had died in the general case since you can do `Duktape.Thread.yield(undefined)`.  ES6 differentiates the two cases by actually returning an object `{ done: true/false, value: xxx }`, so to mimic that I still need to know the alive state... :)

**Comment by svaarala** [comment]
Yeah I know - just meant that if that behavior is a better default it can be changed. That's separate from knowing the state, I understand that.

**Comment by fatcerberus** [comment]
Hm, now that I look a bit more closely, ES6 generator instances actually have the exact same API as iterators.  So any work towards adding generator support to Duktape would also be a move towards adding iterator support :)

**Comment by svaarala** [comment]
So, if you polyfill for `isThreadFinished()` as `Duktape.info(thread).tstate === 5`, would that be enough to polyfill for generator yield/resume?

One intermediate step towards generator support would be to add `Duktape.Thread.prototype.next()` which would have ES6 compatible behavior. `Duktape.Thread.resume()` would still remain; it'd be a bit more lightweight because the return value is directly the return value, and it would also allow resuming with an error.

I'm not sure if this is worth it - but if it had a 90% chance of going in the right direction, it would be quite easy to add.

**Comment by fatcerberus** [comment]
I think so.  Given the thread state is exposed, the code below would then construct a function which behaves exactly like a generator:

```js
function makeGen(fn) {
	return function()
	{
		var thr = new Duktape.Thread(fn);
		return { next: next };
		
		function next()
		{
			var value = Duktape.Thread.resume(thr);  // assuming no-throw behavior
			var done = Duktape.info(thread).tstate === 5;
			return { value: value, done: done };
		}
	};
}

var myGenFunc = makeGen(function()
{
    for (var i = 1; i <= 10; ++i)
       Duktape.Thread.yield(i);
    return "all done!";
});

// ES6 behavior from here on out
var iter = myGenFunc();
var result;
while (!(result = iter.next()).done)
    console.log(result.value);
console.log(result.value);
```

The only thing missing is the syntax sugar (`yield` statement, `function*` keyword)

**Comment by fatcerberus** [comment]
The part that will be more complicated is the `yield*` statement - i.e. forwarding, "yield from another generator"

**Comment by svaarala** [comment]
The `yield` keyword part would be pretty easy to implement: add an opcode that looks up the current thread, and calls `Duktape.Thread.yield()` for the thread (of course looking up the function safely, using a direct internal reference). The compiler would then accept it in *any* function with no other changes which wouldn't be correct but still maybe a useful step.

**Comment by svaarala** [comment]
As for `yield*` it might be most easily implementable by generating opcodes for an iteration loop similar to a `for-in` enumeration: evaluate the target and implement explicit loop control for resume/yield until the target is exhausted. I'm not sure about the details or if that would match ES6 semantics though

**Comment by svaarala** [comment]
Another approach would be for `yield*` to call into some internal helper function that would explicitly iterate the target from C code. The downside of doing so is that the helper would then be in the callstack while that was in progress. So compiling into a bytecode iteration loop sounds easier to implement.

**Comment by fatcerberus** [comment]
I don't think that second approach would work: There would then be a C function in the callstack, so you wouldn't be able to yield.

**Comment by svaarala** [comment]
That can be worked around in special cases. `Duktape.Thread.yield()` is *already* a C function that yields.

**Comment by svaarala** [comment]
Some more detail: that `Duktape.Thread.yield()` is handled as a longjmp which is caught by the bytecode executor. So it can handle the C activation in the callstack, knowing it came from that function. When the function is later resumed, it gets resumed directly from the bytecode executor. It gets a bit messy though.

So, a `yield*` C handler could do essentially the same thing, but instead of getting the argument directly, it would resume (or iterate) the target for a value. Other than that the mechanics could be the same, provided that when the function is resumed, it will keep on calling that `yield*` handler until the iterator is exhausted.

But backing up a bit, being able to implement `yield` and `yield*` in bytecode would have the advantage of *not* having to resort to longjmp tricks. So that's definitely something to look into IMO.

**Comment by fatcerberus** [comment]
That's true, you would already be *inside* the bytecode executor so it's then just a matter of context switching, the longjmp could be avoided entirely.  Which would be good for performance, too.

**Comment by svaarala** [comment]
I actually considered adding a keyword originally.. But then decided against it, figuring that custom bindings are more palatable than custom keywords ;)

**Comment by fatcerberus** [comment]
Hm, so regarding my polyfill above: I had an idea to use some `.bind.apply()` trickery to let the generator take arguments, but I just discovered this bit, it turns out it's currently not allowed to use a bound function as an entrypoint:
https://github.com/svaarala/duktape/blob/master/src-input/duk_bi_thread.c#L123-L128

So much for that idea :/ (the comment says it can be relaxed, but I wouldn't even know where to start, the longjmps confuse me :)

**Comment by svaarala** [comment]
The longjmps are not very complicated in principle: values are communicated via the longjmp value slots. But there are a lot of cases especially for yield: things propagating out of one thread to another in a cascade etc. The longjmp handlers are also not very cleanly written right now - they're very old code compared to their surroundings. I haven't rewritten them because they might be going away :)

I don't recall off hand why that limitation is in place - I'll check if it's easy to fix.

**Comment by svaarala** [comment]
Hmm, I think one reason may be that if it's a bound function, the ultimate target function must be an Ecmascript function. And the code would need to check that to be sure.

**Comment by svaarala** [comment]
Right, and the other issue is that the call setup code expects that the target function can be handled using an "Ecmascript-to-Ecmascript" call setup so that we don't need to do C recursion. That should be now OK for call handling.

**Comment by fatcerberus** [comment]
So another complication came up: It's not currently possible to implement `Generator.prototype.throw()` (i.e. resume with error) or `Generator.prototype.return()` (essentially injecting a `return` statement).  I don't care much about those for the purposes of my polyfill, but for ES6 the necessary mechanism will eventually need to be implemented.

**Comment by Type1J** [comment]
Please continue this conversation! I'd love to see generators in DukTape soon.

**Comment by fatcerberus** [comment]
@Type1J I plan to look into implementing some kind of generator support soon, sometime after Duktape 2.0.0 is released.  It's a feature I'd like to see too :)

**Comment by fatcerberus** [comment]
I had an idea the other day that Duktape threads may be a bit heavyweight (longjmp is a performance concern, etc.) for use as ES6 generators and that the feature could maybe be implemented as "resumable calls": When the function `yield`s it acts like a return, but Duktape remembers the PC within the function as well as any local variables + its scope chain.  Then when you call `next()` it resumes at that point.  I think that could be implemented without a longjmp?

**Comment by svaarala** [comment]
That's basically how Duktape threads already work: they retain their Ecmascript callstack and remember the PC in each call, so that they can be resumed.

Longjmp is by no means required to implement resume/yield. It's just the current implementation because resume() and yield() are functions rather than keywords or opcodes. But even if they are functions it would be possible to avoid the longjmp if the executor special cased the resume/yield calls like it does for eval.

**Comment by svaarala** [comment]
Some more detail on above: so what the executor would like to do is to resolve the Duktape.Thread.resume() and Duktape.Thread.yield() calls and detect that they in fact resolve to the exact native functions implementing a yield/resume. This needs to be done so that if the bindings are overwritten, that also works correctly, so you can't just assume during compilation that Duktape.Thread.resume() would actually be a resume.

Having done that, the executor would just skip the native call altogether and handle the yield/resume as if it was a keyword. Longjmp wouldn't be required.

Another option would be to call the yield/resume function but allow it to return something special so that the yield/resume mechanics could happen without a longjmp. This could be based on a special Duktape/C function return code, or again detecting the exact native function and treating its return value specially.

Direct eval calls already have a special pre-step because direct and indirect eval are handled differently (but even a direct-looking eval("xxx") might not actually be eval() if it has been overwritten) based on the resolved native function pointer.

I'd like to avoid comparing a bunch of function pointers because it affects each call, so ideally a `duk_hobject` could have a flag for "special function call handling". Native function special behaviors would only be checked if the flag was set, and it would only be set for eval, yield, and resume. Right now there isn't a free flag field I think so first one would need to make some room for it.

**Comment by svaarala** [comment]
As a TL;DR summary, I don't think that the yield/resume longjmp() implementation should be a blocker for generators. It can be reworked separately and would benefit yield/resume and generators.

**Comment by fatcerberus** [comment]
One benefit for generators is that `yield` is a keyword.  So no native call is needed there.  *Resuming* is still an issue though: As per the ES iterator protocol, you must call `iter.next()` (or `iter.throw()`).

Agree that it's not the blocker if what you said is true - for some reason I thought the longjmp was integral to Duktape's coroutine implementation.  Good to hear it's not. :)

**Comment by svaarala** [comment]
It's true - I've already prototyped it once but I'd like to avoid doing a lot of function comparisons (well, 3 at present) for every other call.

**Comment by svaarala** [comment]
For generators, .next() and .throw() would be handled like Duktape.Thread.resume() I think - whatever avoids longjmp in the latter would also work for the former.

**Comment by svaarala** [comment]
One relatively simple approach would be to flag each special function (Duktape.Thread.resume/yield, iter.next(), iter.throw(), eval()) with a duk_hobject flag, and then use their magic values as a switch-case argument for special handling. They would in effect be "keywords" (magic would indicate the specific case). Dispatch would need one bit check and one switch-case, which would extend more nicely than doing a growing number of function pointer comparisons.

**Comment by svaarala** [comment]
As for concrete steps forward, there are many alternatives for that. I could try to rework the current yield/resume to be longjmp free first, keeping an eye on making the changes align well with generator requirements. Or generator work could just use longjmp and probably duplicate yield/resume a bit. I'd probably prefer the former because it minimizes overall work but it'll take a bit of time before I'll have time to work on that: right now I'm reworking try-catch handling which is also longjmp related so I'd like to finish that first. But working on generators with longjmps would also be fine - for that it'd be nice to have some idea of what changes and what pulls would be needed for an initial minimal implementation (not necessarily ES2015 compliant but something to start from).

**Comment by svaarala** [comment]
Just checked the duk_hobject flags and the easiest flag to free up would probably be `DUK_HOBJECT_FLAG_EXOTIC_DUKFUNC`, which would be freed if the virtual `.length` was inherited from a NativeFunction prototype. Flags are tight :-)
